File: application/commands/string.go
// Copyright (C) 2019-2026 Ni Rui <ranqus@gmail.com>
// Copyright (C) 2026 Snuffy2
// SPDX-License-Identifier: AGPL-3.0-only

package commands

import (
	"errors"

	"github.com/Snuffy2/sshwifty/application/rw"
)

// Errors
var (
	ErrStringParseBufferTooSmall = errors.New(
		"not enough buffer space to parse given string")

	ErrStringMarshalBufferTooSmall = errors.New(
		"not enough buffer space to marshal given string")
)

// String is a length-prefixed byte string used in the command wire protocol.
// The length field is stored as a variable-length Integer; data is the raw
// byte payload.
type String struct {
	// len is the encoded length of the data payload.
	len Integer
	// data holds the string bytes, aliased into the caller's buffer.
	data []byte
}

// ParseString decodes a length-prefixed String from reader into the provided
// scratch buffer b. It returns ErrStringParseBufferTooSmall if b cannot hold
// the declared payload length.
func ParseString(reader rw.ReaderFunc, b []byte) (String, error) {
	lenData := Integer(0)

	mErr := lenData.Unmarshal(reader)

	if mErr != nil {
		return String{}, mErr
	}

	bLen := len(b)

	if bLen < lenData.Int() {
		return String{}, ErrStringParseBufferTooSmall
	}

	_, rErr := rw.ReadFull(reader, b[:lenData])

	if rErr != nil {
		return String{}, rErr
	}

	return String{
		len:  lenData,
		data: b[:lenData],
	}, nil
}

// NewString constructs a String from raw bytes d. It panics if len(d) exceeds
// MaxInteger, which is the maximum encodable length.
func NewString(d []byte) String {
	dLen := len(d)

	if dLen > MaxInteger {
		panic("Data was too long for a String")
	}

	return String{
		len:  Integer(dLen),
		data: d,
	}
}

// Data returns the data of the string
func (s String) Data() []byte {
	return s.data
}

// Marshal encodes the String into b, writing the variable-length length prefix
// followed by the data bytes. It returns ErrStringMarshalBufferTooSmall if b
// is not large enough.
func (s String) Marshal(b []byte) (int, error) {
	bLen := len(b)

	if bLen < s.len.ByteSize()+len(s.data) {
		return 0, ErrStringMarshalBufferTooSmall
	}

	mLen, mErr := s.len.Marshal(b)

	if mErr != nil {
		return 0, mErr
	}

	return copy(b[mLen:], s.data) + mLen, nil
}

File: ui/commands/string.js
// Copyright (C) 2019-2026 Ni Rui <ranqus@gmail.com>
// Copyright (C) 2026 Snuffy2
// SPDX-License-Identifier: AGPL-3.0-only

/**
 * @file Length-prefixed string encoding for the ShellPort command protocol.
 *
 * {@link String} wraps a raw `Uint8Array` and can serialize itself as a
 * variable-length integer length prefix followed by the string bytes
 * ({@link String#buffer}), or decode itself from a stream reader
 * ({@link String.read}). The {@link truncate} helper is provided for capping
 * display strings in the UI.
 */

import * as reader from "../stream/reader.js";
import * as integer from "./integer.js";

/**
 * Length-prefixed binary string for the command protocol.
 *
 * Wire format: `[Integer length][...bytes]`. The static {@link String.read}
 * factory decodes from a stream reader; {@link String#buffer} encodes for
 * sending.
 */
export class String {
  /**
   * Read String from given reader
   *
   * @param {reader.Reader} rd Source reader
   *
   * @returns {String} readed string
   *
   */
  static async read(rd) {
    let l = new integer.Integer(0);

    await l.unmarshal(rd);

    return new String(await reader.readN(rd, l.value()));
  }

  /**
   * constructor
   *
   * @param {Uint8Array} str String data
   */
  constructor(str) {
    this.str = str;
  }

  /**
   * Return the string
   *
   * @returns {Uint8Array} String data
   *
   */
  data() {
    return this.str;
  }

  /**
   * Return serialized String as array
   *
   * @returns {Uint8Array} serialized String
   *
   */
  buffer() {
    let lBytes = new integer.Integer(this.str.length).marshal(),
      buf = new Uint8Array(lBytes.length + this.str.length);

    buf.set(lBytes, 0);
    buf.set(this.str, lBytes.length);

    return buf;
  }
}

/**
 * Truncates a string to the maximum length
 *
 * @param {string} str Source string
 * @param {integer} maxLength Max length
 * @param {string} exceed Text appends the string if it was truncated
 *
 * @returns {string} truncated String
 *
 */
export function truncate(str, maxLength, exceed) {
  if (str.length <= maxLength) {
    return str;
  }
  return str.substring(0, maxLength) + exceed;
}
